### ISSUE TITLE
Extend support to running on Windows

### ISSUE THREAD

**Issue by mbwhite** [hyperledger/composer#65]
Extend support for Fabric-Composer onto Windows platforms. 

---
**NOTE:** Windows is not currently supported for Composer due to issues documented below. We would like to support it so if you can help to resolve the issue below it would be appreciated.

Windows 10 Professional will be the minimum version of Windows.

If Windows is your desktop platform, then we would sugget using something Vagrant to create a VM that you can deploy Composer/Fabric on (see [https://github.com/davidkel/ComposerVagrant](https://github.com/davidkel/ComposerVagrant))

**Comment by sstone1** [comment]
We cannot do this without setting up some form of continuous integration, looks like Appveyor would do the job: https://www.appveyor.com/docs/lang/nodejs-iojs/

**Comment by mbwhite** [comment]
# Results so far

Options for running Fabric and Fabric Composer on Windows

## Option 1:
Use a virtual machine such as OpenBox to run a Ubuntu image in and develop the applications and run HFC/Composer there.  That does work well (assuming you have a laptop with the memory and a disk speed)

## Option 2:
Similar to option 1 - but use Vagrant to create a Linux vmto run the HFC fabric images in (via docker) and the connect to it from applications in windows / or edit in Windows but on shared files running in the vm

## Option 3:
On Windows 10 Pro
 - enable the Hyper-V support and install docker natively
 - Install the Windows GIT client & shell
 - Install the Windows Node.js runtime
 - For the native modules install https://www.npmjs.com/package/windows-build-tools

You can then proceed as per normal. The git bash shell is very useful to run the scripts provided with windows. But you can just issue the same commands from a cmd.exe shell - just the scripts as such don't work. Note `npm run <scriptname>` doesn't work even in the git bash shell

*There are problems*
 - The directory c:\tmp needs to exist  --> https://github.com/IBM-Blockchain/cp-web/issues/56  
 - Deploying the business network is resulting in a segmentation fault from node.js

_Need to dig further into what the issue is here_

## Option 4:
Use the Windows 10 beta bash shell.. initial results where promising here - but the firewall I had at the time was the wrong version for windows 10 so was making this too hard to be practical. It asked for confirmation on almost every tcp/ip connection.. Worth trying again.

**Comment by mbwhite** [comment]
Diagnostics copied from windbg
[node.js error.txt](https://github.com/fabric-composer/fabric-composer/files/864832/node.js.error.txt)

**Comment by mbwhite** [comment]
In addition - if you can get Docker running on Windows there is a prebuilt version of the Playground that is available https://hub.docker.com/r/fabriccomposer/composer-playground/

**Comment by mbwhite** [comment]
I've been able to make significant progress on this; the feasibility of using Windows 10 Professional as a **development** platform is plausible. 

**Pre-preqs**

- Windows 10 Pro, with hyper-v enabled
- Docker for Windows installed
- Node 6 and npm  4 installed
  - The version of npm with Node is out of date and should be updated `npm install npm@latest --global`
- Git for windows, installed with the option of adding the git bash tools added to the path for all users
  - _this is needed as currently all the scripts are setup as bash scripts, they could be converted to Windows shell scripts_

**Essential setup**

I had to create a new local user in the Windows command prompt, this was needed as due to to the AzureDomain login of my laptop I needed another user to give the correct credentials to Docker.

Docker Shared Folders has to be enabled as the Fabric Docker containers are Linux. 

**Fixes that are needed**

1. The `composer archive create` needs a fix as the transaction function javascript has the wrong path
2. There is a fix that is needed in the fabric-client, Golang.js file, findSource() method.

```
  var desc = {
	name: path.relative(goPath, entry.path).replace(/\\/g, '/'),
	fqp: entry.path.replace(/\\/g, '/')
};
```
3. I had to replace the hastable node module with the hashtable-js module (drop in replacement) in the EventHub.js file. hastable-js is a pure javascript version of the native C++ code in the hastable module. This was generating Access Violations when a network was being deployed

**General Fixes**

The npm install on all modules had to be done as the Windows admin user, as that was the only one that could see the c compiler. And some care had to be taken to ensure that all the native libraries where built. However I think this is a path issue.

All the scripts in the Composer repos are based on bash, therefore the git tool was very useful for running these. 

**Conclusion**
This is plausible, however a discussion about how much Windows-centric the scripts etc should be. 

Running the Composer-Playground all-in-one docker image runs without issue - though that does need the same separate user approach potentially.

**Comment by mbwhite** [comment]
Regarding the windows build tools, that are very useful for getting native code to compile - they can be tricky to install if your user ID is not an administrator. 

As in the comment above I was using a new local userid for running the Docker containers - as I was unable to use my usual logon due to AzureAD Enterprise setup. Originally I had bootstrapped the example by using my logon for the `npm install`;

It's worth reviewing this[ issue in github](https://github.com/felixrieseberg/windows-build-tools/issues/54) for more information. This shows how you can install the build tools from an administor (elevated) command prompt into specific users directory (in the case above *lennyfabric*)

In summary the process should be to create an elevated command prompt, then apply the settings as described in the git issue above, and do the installation of the windows build tools. This will then install the tools into the user of your choice ( *lennyfabric* in the example above).

In the examples then the npm install can proceed as needed.

**Comment by mbwhite** [comment]
The issue for the windows paths has been resolved by the fabric issue = https://gerrit.hyperledger.org/r/#/c/9467/

The crash can be resolved by https://jira.hyperledger.org/browse/FAB-3977

Within the `composer archive create` there was an error re the paths as well. This has been resolved and will be in the next release.

**Comment by davidkel** [comment]
With the beta release, it looks like a new native dependency has been added. node-x509 (listed just as x509). This looks to be pulled in by fabric-ca-client node module.
This makes it unpractical to use npm install for anyone but advanced windows developers to resolve at this time.

**Comment by naveensachdeva** [comment]
I successfully installed Hyperledger Composer Playground locally on my Windows 7 machine but running into an issue.

The install was successful and I am able to access the Playground.
The install also created a connection profile (which looks to be the correct connection profile) in .composer-connection-profiles/hlfv1 folder, along with PeerAdmin identity in .hfc-key-store folder. 
But the install did not set/imported this connection profile in Playground.
I was able to import this connection profile, but when I try to "Use this profile" and connect using the PeerAdmin identity, I get - "Failed to connect: Error: error trying to setup a keystore path. EACCES" :(

If it is any help, I see the following in the browser console -
Error: error trying to setup a keystore path. EACCES
    at Client.newDefaultKeyValueStore.then.catch (/usr/local/lib/node_modules/composer-playground/node_modules/composer-connector-hlfv1/lib/hlfconnectionmanager.js:339:36)
    at process._tickDomainCallback (internal/process/next_tick.js:135:7)

**Comment by mbwhite** [comment]
@naveensachdeva  thank you for your comment - however at this point in time we would not support or recommended using Composer on Window; this is not so we don't want to but there are restrictions as documented above in the Fabric libraries that prevent this.  We want to get full end-end support.

Windows 10 also will be the minimum due to the Docker / HyperV support that is required for Fabric. 

If Windows is your desktop platform, then we would sugget using something Vagrant to create a VM that you can deploy Composer/Fabric on (see https://github.com/davidkel/ComposerVagrant)

**Comment by davidkel** [comment]
#1628 Has some details of where I have got so far, it is looking very promising, I just need to get a Windows 10 Pro system to be able to run docker for windows

**Comment by im89097733** [comment]
Hello:  

I am not sure where the Win 10 support is at.  I see @davidkel posted a few days ago.
I seem to be very close with the curl from https://hyperledger.github.io/composer/installing/using-playground-locally.html

I have Win 10 Pro with HyperV setup and Docker working per requirements.
The error I get is with a failure to find the composer-channel.tx file.  it does not exist in the portable git dir.
I suspected a permissions problem as a reason why the config is not getting created and/or mapped into the container.

@mbwhite do I need to create a separate user?  I was about to try running my shell as admin 

BTW: not opposed to running  linux variant VM but would rather not and if I can help with testing please LMK

**Comment by mbwhite** [comment]
@im89097733 thank you for the offer; certainly help with testing would be appreciated. 
@davidkel will be able to confirm/deny the exact state. 

As to the separate user.. I certainly had to create one to get docker to work properly; however this is primarily related to the Enterprise configuration I have on my laptop;

**Comment by davidkel** [comment]
@im89097733 I was thinking about getting it working on Windows 10 Pro Creator release using WSL. My initial comments can be found here #1628 
I can certainly get the CLI/Playground working without any problems and would be able to connect to a remote fabric, I wanted to confirm I could control a Fabric deployed to Docker for Windows (the hyperv version) using docker-compose and docker from WSL. 
Once confirmed then we would look at how to automate delivering our fabric-dev-server so it could run locally.
Following that we would then need to look at the 1-click install scripts to see how they fare.

**Comment by knagaev** [comment]
I inserted in startFabric.sh

`export MSYS_NO_PATHCONV=1`

before 

```
# Create the channel
docker exec peer0.org1.example.com peer channel create -o orderer.example.com:7050 -c composerchannel -f /etc/hyperledger/configtx/composer-channel.tx
```

and the script runs ok.

**Comment by kletkeman** [comment]
Addition issue with Windows 7 and 10: Just cloning a sample project fails with many file paths being too long. I have a BNA that works great on the Mac (and presumably Linux, which I am moving to instead of Windows), but while it clones ok on Windows, it cannot check out. The following errors come from Github Desktop native (same UI I use on Mac, which works there):

Cloning into 'D:\My Documents\github.ibm.com\kletkema\iot-blockchain'...
remote: Counting objects: 7095, done.        
remote: Compressing objects:   0% (1/126)           
__... cut for brevity ...__
Receiving objects:  93% (6599/7095), 2.39 MiB | 4.73 MiB/s   
Receiving objects:  94% (6670/7095), 2.39 MiB | 4.73 MiB/s   
Receiving objects:  95% (6741/7095), 2.39 MiB | 4.73 MiB/s   
Receiving objects:  96% (6812/7095), 2.39 MiB | 4.73 MiB/s   
Receiving objects:  97% (6883/7095), 2.39 MiB | 4.73 MiB/s   
Receiving objects:  98% (6954/7095), 2.39 MiB | 4.73 MiB/s   
Receiving objects:  99% (7025/7095), 2.39 MiB | 4.73 MiB/s   
remote: Total 7095 (delta 17), reused 0 (delta 0), pack-reused 6952        
Receiving objects: 100% (7095/7095), 2.39 MiB | 4.73 MiB/s   
Receiving objects: 100% (7095/7095), 3.35 MiB | 5.17 MiB/s, done.
Resolving deltas:   0% (0/3841)   
Resolving deltas:   1% (45/3841)   
__... cut for brevity ...__
Resolving deltas: 100% (3841/3841)   
Resolving deltas: 100% (3841/3841), done.
error: unable to create file samples/maximo_poc/composer-data/fabric-dev-servers/fabric-scripts/hlfv1/composer/crypto-config/ordererOrganizations/example.com/orderers/orderer.example.com/msp/keystore/4947ca37ff6aca98aed30b2f50b64b152f22089e51cf0a70ff5ceb76e9872d66_sk: Filename too long
error: unable to create file samples/maximo_poc/composer-data/fabric-dev-servers/fabric-scripts/hlfv1/composer/crypto-config/ordererOrganizations/example.com/users/Admin@example.com/msp/keystore/c6211e0b87d5ac94276dbf92e4cfadf385ee78e1c46cffb9a2e454e090736065_sk: Filename too long
error: unable to create file samples/maximo_poc/composer-data/fabric-dev-servers/fabric-scripts/hlfv1/composer/crypto-config/peerOrganizations/org1.example.com/peers/peer0.org1.example.com/msp/admincerts/Admin@org1.example.com-cert.pem: Filename too long
error: unable to create file samples/maximo_poc/composer-data/fabric-dev-servers/fabric-scripts/hlfv1/composer/crypto-config/peerOrganizations/org1.example.com/peers/peer0.org1.example.com/msp/keystore/dfb17cf51dc061d585b4850599be0e4b8b7cc8cc363a67c23bc03c6c5393b0e0_sk: Filename too long
error: unable to create file samples/maximo_poc/composer-data/fabric-dev-servers/fabric-scripts/hlfv1/composer/crypto-config/peerOrganizations/org1.example.com/peers/peer0.org1.example.com/msp/signcerts/peer0.org1.example.com-cert.pem: Filename too long
error: unable to create file samples/maximo_poc/composer-data/fabric-dev-servers/fabric-scripts/hlfv1/composer/crypto-config/peerOrganizations/org1.example.com/peers/peer0.org1.example.com/msp/tlscacerts/tlsca.org1.example.com-cert.pem: Filename too long
error: unable to create file samples/maximo_poc/composer-data/fabric-dev-servers/fabric-scripts/hlfv1/composer/crypto-config/peerOrganizations/org1.example.com/users/Admin@org1.example.com/msp/admincerts/Admin@org1.example.com-cert.pem: Filename too long
error: unable to create file samples/maximo_poc/composer-data/fabric-dev-servers/fabric-scripts/hlfv1/composer/crypto-config/peerOrganizations/org1.example.com/users/Admin@org1.example.com/msp/keystore/114aab0e76bf0c78308f89efc4b8c9423e31568da0c340ca187a9b17aa9a4457_sk: Filename too long
error: unable to create file samples/maximo_poc/composer-data/fabric-dev-servers/fabric-scripts/hlfv1/composer/crypto-config/peerOrganizations/org1.example.com/users/Admin@org1.example.com/msp/signcerts/Admin@org1.example.com-cert.pem: Filename too long
error: unable to create file samples/maximo_poc/composer-data/fabric-dev-servers/fabric-scripts/hlfv1/composer/crypto-config/peerOrganizations/org1.example.com/users/Admin@org1.example.com/msp/tlscacerts/tlsca.org1.example.com-cert.pem: Filename too long
error: unable to create file samples/maximo_poc/composer-data/fabric-dev-servers/fabric-scripts/hlfv1_Sep17/composer-data/fabric-dev-servers/fabric-scripts/hlfv1/composer/creds/114aab0e76bf0c78308f89efc4b8c9423e31568da0c340ca187a9b17aa9a4457-priv: Filename too long
error: unable to create file samples/maximo_poc/composer-data/fabric-dev-servers/fabric-scripts/hlfv1_Sep17/composer-data/fabric-dev-servers/fabric-scripts/hlfv1/composer/creds/114aab0e76bf0c78308f89efc4b8c9423e31568da0c340ca187a9b17aa9a4457-pub: Filename too long
warning: unable to access 'samples/maximo_poc/composer-data/fabric-dev-servers/fabric-scripts/hlfv1_Sep17/composer-data/fabric-dev-servers/fabric-scripts/hlfv1/composer/crypto-config/ordererOrganizations/example.com/ca/.gitattributes': Filename too long
error: unable to create file samples/maximo_poc/composer-data/fabric-dev-servers/fabric-scripts/hlfv1_Sep17/composer-data/fabric-dev-servers/fabric-scripts/hlfv1/composer/crypto-config/ordererOrganizations/example.com/ca/5a6c4ce688103f8220147c6f8257eefbb5a1a8a80220c720d8152a2158021893_sk: Filename too long
error: unable to create file samples/maximo_poc/composer-data/fabric-dev-servers/fabric-scripts/hlfv1_Sep17/composer-data/fabric-dev-servers/fabric-scripts/hlfv1/composer/crypto-config/ordererOrganizations/example.com/ca/ca.example.com-cert.pem: Filename too long
fatal: cannot create directory at 'samples/maximo_poc/composer-data/fabric-dev-servers/fabric-scripts/hlfv1_Sep17/composer-data/fabric-dev-servers/fabric-scripts/hlfv1/composer/crypto-config/ordererOrganizations/example.com/msp/admincerts': Filename too long
__warning: Clone succeeded, but checkout failed.__
You can inspect what was checked out with 'git status'
and retry the checkout with 'git checkout -f HEAD'

**Comment by davidkel** [comment]
@kletkeman Thanks for the info. This is due to a limitation in windows NTFS. I am assuming that if you used the WSL feature of windows 10 then that would not have this problem (can't believe WSL would enforce the same restriction). Also it seems that for newer editions of windows 10 there is a hidden switch to enable NTFS long filename support, however you need to use the group policy editor to change it, that means it not an option for Windows Home users.

**Comment by davidkel** [comment]
Fabric have now included pkcs11 which requires a compiler. I have requested this be made optional as the majority of composer users will not need this dependency.

**Comment by cupliz** [comment]
i got this problem when instal composer-cli on WSL, may have something to do with pkcs11?
```
gyp WARN EACCES attempting to reinstall using temporary dev dir "/root/.nvm/versions/node/v8.9.3/lib/node_modules/composer-cli/node_modules/pkcs11js/.node-gyp"
```

**Comment by davidkel** [comment]
## native modules currently being included
- pkcs11 from node-sdk. node-sdk will need to make pkcs an optional dependency
- x509 from fabric-shim. This get's pulled in currently because packages depend on composer-runtime-hlfv1. Need to make sure this is fixed as part of native fabric deployment
- node-report is our own inclusion which has a native module, which is a shame.

**Comment by uluhonolulu** [comment]
FYI, I'm running Composer on Windows 7 just fine. Using Docker Tools and the scripts from https://github.com/acloudfan/HLF-Windows-Fabric-Tool (this guy reads a Composer course on Udemy). Worked OOTB for 0.16/1.0, had to modify the corresponding 1.1 scripts for 0.19/1.1.

**Comment by uluhonolulu** [comment]
I've also noticed that the participant is not created right away by calling `await participantRegistry.add(participant);`. I have to poll the registry calling `participantRegistry.exists()` until I get `true`. Doesn't happen on Ubuntu.
